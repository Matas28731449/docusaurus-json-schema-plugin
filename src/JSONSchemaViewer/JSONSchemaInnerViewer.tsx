import React from "react"

import type { JSONSchema7 } from "json-schema"

type InnerViewerProperties = {
  schema: JSONSchema7
}

function JSONSchemaInnerViewer(props: InnerViewerProperties): JSX.Element {
  return <></>
}

export default JSONSchemaInnerViewer
