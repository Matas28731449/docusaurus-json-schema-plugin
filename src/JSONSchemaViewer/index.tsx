import React, { useEffect, useState } from "react"
import Layout from "@theme/Layout"
import Translate from "@docusaurus/Translate"
import AllOfMerger from "json-schema-merge-allof"
import { Resolver } from "@stoplight/json-ref-resolver"

import type { JSONSchema7 } from "json-schema"

import JSONSchemaViewer from "./JSONSchemaInnerViewer"

type ViewerProperties = {
  loadSchema: () => Promise<JSONSchema7>
}

// merged representation of allOf array of schemas
async function mergeAllOf(
  schema: JSONSchema7
): Promise<Omit<JSONSchema7, "allOf">> {
  const mergedSchema = AllOfMerger(schema, {
    // Technically valid, but only interested on fields with name, at least for now ...
    ignoreAdditionalProperties: true,
  })

  return Promise.resolve(mergedSchema as Omit<JSONSchema7, "allOf">)
}

function JSONSchemaEditor(props: ViewerProperties): JSX.Element {
  const [schema, setSchema] = useState(
    undefined as undefined | Omit<JSONSchema7, "allOf">
  )
  const [fetchError, setFetchError] = useState(undefined as undefined | Error)

  useEffect(() => {
    props
      .loadSchema()
      .then((userSchema) => new Resolver().resolve(userSchema))
      .then((resolvedSchema) => mergeAllOf(resolvedSchema.result))
      .then((schemaWithoutAllOf) => setSchema(schemaWithoutAllOf))
      .catch((err) => setFetchError(err))
  }, [])

  return (
    <Layout>
      {fetchError !== undefined && (
        <div>
          <p>
            <Translate
              values={{
                errorMessage: fetchError.message,
                id: "docusaurus-json-schema-viewer-plugin.error",
              }}
            >
              {"This component crashed because of error: {errorMessage}."}
            </Translate>
          </p>
        </div>
      )}
      {schema === undefined && (
        <div>
          <p>
            <Translate values={{ id: "docusaurus-json-schema-viewer-plugin.loading" }}>
              {"Loading schema ..."}
            </Translate>
          </p>
        </div>
      )}
      {schema !== undefined && <JSONSchemaViewer schema={schema} />}
    </Layout>
  )
}

export default JSONSchemaEditor
