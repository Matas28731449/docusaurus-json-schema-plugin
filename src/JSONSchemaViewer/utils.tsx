import React from "react";

import type { JSONSchema7 } from "json-schema";

// To keep the same type everywhere here
type cleanSchema = Omit<JSONSchema7, "allOf">;

