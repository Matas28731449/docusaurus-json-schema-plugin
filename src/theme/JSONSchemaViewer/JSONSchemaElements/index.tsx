// Component to deal with oneOf / anyOf / allOf / not
export { SchemaComposition } from "./schemaComposition/index"

// Component to deal with if-then-else , dependentRequired , dependentSchemas , dependencies
export { SchemaConditional } from "./SchemaConditional/index"

// Component to deal with Object
export { CreateObject } from "./object/index"

// Component to deal with Array
export { CreateArray } from "./array/index"

// Component to deal with String
export { default as CreateString } from "./CreateString"

// Component to deal with Boolean
export { default as CreateBoolean } from "./CreateBoolean"

// Component to deal with Number
export { default as CreateNumber } from "./CreateNumber"

// Component to deal with Integer
export { default as CreateInteger } from "./CreateInteger"

// Component to deal with schema defined as "true"
export { default as CreateAlwaysValid } from "./CreateAlwaysValid"

// Component to deal with schema defined as "false"
export { default as CreateAlwaysInvalid } from "./CreateAlwaysInvalid"
