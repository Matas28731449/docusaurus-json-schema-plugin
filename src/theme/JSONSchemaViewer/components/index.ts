// To avoid circular dependancies issue, let's have a single place for exports and imports

// Collapsible componenent
export { default as Collapsible } from "./Collapsible"

// CreateDetailsNode component
export { default as CreateDetailsNode } from "./CreateDetailsNode"

// CreateEdges component
export { default as CreateEdges } from "./CreateEdges"

// CreateProperties component
export { default as CreateProperties } from "./CreateProperties"

// RenderAnyOneOf component
export { default as RenderAnyOneOf } from "./RenderAnyOneOf"

// https://stackoverflow.com/a/69328045/6149867
export type WithRequired<T, K extends keyof T> = T & { [P in K]-?: T[P] }
