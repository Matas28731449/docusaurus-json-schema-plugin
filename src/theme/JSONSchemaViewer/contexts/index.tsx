export * from "@theme/JSONSchemaViewer/contexts/jsvOptions"
export * from "@theme/JSONSchemaViewer/contexts/schemaHierarchy"
