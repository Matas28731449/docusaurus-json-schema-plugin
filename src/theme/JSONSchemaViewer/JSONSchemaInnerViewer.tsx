import React from "react"
import { CreateNodes } from "./components/index"

import type { JSONSchema7 } from "json-schema"

type InnerViewerProperties = {
  // Thanks to json-schema-merge-allof , we don't have allOf in the whole user schema
  // Thanks to @stoplight/json-ref-resolver, $ref are either :
  // 1. resolved
  // 2. unresolved (as circular stuff are not on the roadmap)
  schema: Omit<JSONSchema7, "allOf">
}

// External
function JSONSchemaInnerViewer(props: InnerViewerProperties): JSX.Element {
  return <ul style={{ marginLeft: "1rem" }}>{CreateNodes(props.schema)}</ul>
}

export default JSONSchemaInnerViewer
