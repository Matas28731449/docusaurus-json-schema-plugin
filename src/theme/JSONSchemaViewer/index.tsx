import React, { useState, useEffect } from "react"
import { Resolver } from "@stoplight/json-ref-resolver"

import { CreateNodes, Collapsible } from "./components/index"
import { JSVOptionsContextProvider } from "./contexts/index"

import type { JSONSchema } from "./types"
import type { JSVOptions } from "./contexts/index"
import type { IResolveOpts } from "@stoplight/json-ref-resolver/types"

export type Props = {
  // The schema to use
  schema: unknown
  // To customize the ref resolving
  resolverOptions?: IResolveOpts
  // To customize the viewer
  viewerOptions?: Omit<JSVOptions, "fullSchema">
  // Other properties (who knows ?)
  [x: string]: any
}

type InnerViewerProperties = {
  // Thanks to @stoplight/json-ref-resolver, $ref are either :
  // 1. resolved
  // 2. unresolved (as circular stuff are not on the roadmap)
  schema: JSONSchema
  // Options for viewer
  viewerOptions?: Omit<JSVOptions, "fullSchema">
}

// Internal
function JSONSchemaInnerViewer(props: InnerViewerProperties): JSX.Element {
  const { schema, viewerOptions } = props
  // Title of the schema, for user friendliness
  const title =
    typeof schema !== "boolean" && schema?.title !== undefined
      ? schema.title
      : "Schema"

  // state for provider
  const startingState: JSVOptions = {
    fullSchema: schema,
    // spread provided attributes
    ...viewerOptions,
  }

  return (
    <JSVOptionsContextProvider value={startingState}>
      <Collapsible
        summary={<strong>{title}</strong>}
        detailsProps={{
          open: true,
        }}
      >
        <CreateNodes schema={schema} />
      </Collapsible>
    </JSVOptionsContextProvider>
  )
}

// Entry point
export default function JSONSchemaViewer(props: Props): JSX.Element {
  const { schema: originalSchema, resolverOptions, viewerOptions } = props

  const [error, setError] = useState(undefined as undefined | Error)
  const [resolvedSchema, setResolvedSchema] = useState(
    undefined as undefined | JSONSchema
  )

  useEffect(() => {
    // Time to do the job
    new Resolver()
      .resolve(originalSchema, resolverOptions)
      .then((result) => {
        setResolvedSchema(result.result)
      })
      .catch((err) => {
        setError(err)
      })
  }, [])

  if (error !== undefined) {
    return <div>Something bad happens : {error.message}</div>
  } else if (resolvedSchema === undefined) {
    return <div>Loading ....</div>
  } else {
    return (
      <JSONSchemaInnerViewer
        schema={resolvedSchema}
        viewerOptions={viewerOptions}
      />
    )
  }
}
