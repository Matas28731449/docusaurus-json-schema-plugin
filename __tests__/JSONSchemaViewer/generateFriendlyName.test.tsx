import React from "react"

// For typings autocomplete whatever your IDE
import { expect, test, describe } from "@jest/globals"

import { create, act } from "react-test-renderer"

import JSONSchemaViewer from "../../src/theme/JSONSchemaViewer/index"

// Type to prevent creating invalid mocks
import type {
  JSONSchema,
  JSONSchemaNS,
} from "../../src/theme/JSONSchemaViewer/types"

// Type for react-test-renderer
import type { ReactTestRenderer } from "react-test-renderer"

const testcases: [string, JSONSchema][] = [
  [
    "Prefer title",
    {
      type: "array",
      prefixItems: [
        {
          title: "My custom title",
          type: "object",
          properties: {
            text: {
              const: "Hello World",
            },
          },
        },
      ],
    } as JSONSchemaNS.Array,
  ],
  [
    "Number without explicit type",
    {
      type: "array",
      prefixItems: [
        {
          multipleOf: 10,
        },
      ],
    } as JSONSchemaNS.Array,
  ],
  [
    "Object without explicit type",
    {
      type: "array",
      prefixItems: [
        {
          minProperties: 3,
        },
      ],
    } as JSONSchemaNS.Array,
  ],
  [
    "Multiple types",
    {
      type: "array",
      prefixItems: [
        {
          type: ["null", "number"],
        },
      ],
    } as JSONSchemaNS.Array,
  ],
  [
    "Not (integer)",
    {
      type: "array",
      prefixItems: [
        {
          not: {
            type: "integer",
          },
        },
      ],
    } as JSONSchemaNS.Array,
  ],
  [
    "OR / XOR / AND",
    {
      type: "array",
      prefixItems: [
        {
          anyOf: [
            { type: "string", maxLength: 5 },
            { type: "number", minimum: 0 },
          ],
        },
        {
          oneOf: [
            { type: "number", multipleOf: 5 },
            { type: "number", multipleOf: 3 },
          ],
        },
        {
          allOf: [{ type: "string" }, { maxLength: 5 }],
        },
      ],
    } as JSONSchemaNS.Array,
  ],
]

describe("JSONSchemaViewer - generateFriendlyName cases", () => {
  test.each(testcases)("test %s", async (_title, fakeSchema) => {
    // render the component
    let root: ReactTestRenderer | undefined

    await act(async () => {
      root = create(<JSONSchemaViewer schema={fakeSchema} />)
    })

    // make assertions on root
    expect(root?.toJSON()).toMatchSnapshot()
  })
})
