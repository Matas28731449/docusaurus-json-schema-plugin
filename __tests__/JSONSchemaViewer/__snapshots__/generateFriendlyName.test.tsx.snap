// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONSchemaViewer - generateFriendlyName cases test Array specified only with items 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <ul>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items[0]
          </code>
          Â 
          <span>
            (integer OR boolean)[]
          </span>
        </summary>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style={
            {
              "opacity": "0.6",
            }
          }
        >
          array
        </span>
        <ul>
          <li>
            <details
              open={false}
            >
              <summary>
                <code>
                  items
                </code>
                Â 
                <span>
                  integer OR boolean
                </span>
              </summary>
              <div
                className="tabs-container"
              >
                <ul
                  aria-orientation="horizontal"
                  className="tabs"
                  role="tablist"
                >
                  <li
                    aria-selected={true}
                    onClick={[Function]}
                    tabIndex={0}
                  >
                    <span
                      style={
                        {
                          "opacity": "0.6",
                        }
                      }
                    >
                      integer
                    </span>
                  </li>
                  <li
                    aria-selected={false}
                    onClick={[Function]}
                    tabIndex={-1}
                  >
                    <span
                      style={
                        {
                          "opacity": "0.6",
                        }
                      }
                    >
                      boolean
                    </span>
                  </li>
                </ul>
                <div>
                  <div
                    hidden={false}
                    role="tabpanel"
                  >
                    <div
                      role="tabpanel"
                    >
                      <strong>
                        type
                      </strong>
                      Â :Â 
                      <span
                        style={
                          {
                            "opacity": "0.6",
                          }
                        }
                      >
                        integer
                      </span>
                      <div
                        style={
                          {
                            "marginTop": "var(--ifm-table-cell-padding)",
                          }
                        }
                      />
                    </div>
                  </div>
                  <div
                    hidden={true}
                    role="tabpanel"
                  >
                    <div
                      role="tabpanel"
                    >
                      <strong>
                        type
                      </strong>
                      Â :Â 
                      <span
                        style={
                          {
                            "opacity": "0.6",
                          }
                        }
                      >
                        boolean
                      </span>
                      <div
                        style={
                          {
                            "marginTop": "var(--ifm-table-cell-padding)",
                          }
                        }
                      />
                    </div>
                  </div>
                </div>
              </div>
            </details>
          </li>
        </ul>
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        >
          <div>
            ðŸš¨Â 
            No extra item(s) are authorized in this array
          </div>
        </div>
      </details>
    </li>
  </ul>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  />
</details>
`;

exports[`JSONSchemaViewer - generateFriendlyName cases test Array specified only with prefixItems 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <ul>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items[0]
          </code>
          Â 
          <span>
            (boolean,integer)[]
          </span>
        </summary>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style={
            {
              "opacity": "0.6",
            }
          }
        >
          array
        </span>
        <ul />
        <ul>
          <li>
            <details
              open={false}
            >
              <summary>
                <code>
                  items[0]
                </code>
                Â 
                <span>
                  boolean
                </span>
              </summary>
              <strong>
                type
              </strong>
              Â :Â 
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                boolean
              </span>
              <div
                style={
                  {
                    "marginTop": "var(--ifm-table-cell-padding)",
                  }
                }
              />
            </details>
          </li>
          <li>
            <details
              open={false}
            >
              <summary>
                <code>
                  items[1]
                </code>
                Â 
                <span>
                  integer
                </span>
              </summary>
              <strong>
                type
              </strong>
              Â :Â 
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                integer
              </span>
              <div
                style={
                  {
                    "marginTop": "var(--ifm-table-cell-padding)",
                  }
                }
              />
            </details>
          </li>
        </ul>
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        >
          <div>
            ðŸš¨Â 
            No extra item(s) are authorized in this array
          </div>
        </div>
      </details>
    </li>
  </ul>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  />
</details>
`;

exports[`JSONSchemaViewer - generateFriendlyName cases test Array specified with items and contains 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <ul>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items[0]
          </code>
          Â 
          <span>
            (number,...,string,...)[]
          </span>
        </summary>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style={
            {
              "opacity": "0.6",
            }
          }
        >
          array
        </span>
        <ul>
          <li>
            <details
              open={false}
            >
              <summary>
                <code>
                  items[0]
                </code>
                Â 
                <span>
                  number
                </span>
              </summary>
              <strong>
                type
              </strong>
              Â :Â 
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                number
              </span>
              <div
                style={
                  {
                    "marginTop": "var(--ifm-table-cell-padding)",
                  }
                }
              />
            </details>
          </li>
        </ul>
        <ul>
          <li>
            <details
              open={false}
            >
              <summary>
                <code>
                  contains
                </code>
                Â 
                <span>
                  string
                </span>
              </summary>
              <strong>
                type
              </strong>
              Â :Â 
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                string
              </span>
              <div
                style={
                  {
                    "marginTop": "var(--ifm-table-cell-padding)",
                  }
                }
              />
            </details>
          </li>
        </ul>
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        />
      </details>
    </li>
  </ul>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  />
</details>
`;

exports[`JSONSchemaViewer - generateFriendlyName cases test Multiple types 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <ul>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items[0]
          </code>
          Â 
          <span>
            number
          </span>
        </summary>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style={
            {
              "opacity": "0.6",
            }
          }
        >
          number
        </span>
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        />
      </details>
    </li>
  </ul>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  />
</details>
`;

exports[`JSONSchemaViewer - generateFriendlyName cases test Not (integer) 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <ul>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items[0]
          </code>
          Â 
          <span>
            NOT (integer)
          </span>
        </summary>
        <div>
          <span
            className="badge badge--danger"
          >
            not
          </span>
          <br />
          <strong>
            type
          </strong>
          Â :Â 
          <span
            style={
              {
                "opacity": "0.6",
              }
            }
          >
            integer
          </span>
          <div
            style={
              {
                "marginTop": "var(--ifm-table-cell-padding)",
              }
            }
          />
        </div>
      </details>
    </li>
  </ul>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  />
</details>
`;

exports[`JSONSchemaViewer - generateFriendlyName cases test Number without explicit type 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <ul>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items[0]
          </code>
          Â 
          <span>
            number
          </span>
        </summary>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style={
            {
              "opacity": "0.6",
            }
          }
        >
          number
        </span>
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        >
          <div>
            <strong>
              Possible values :
            </strong>
            Â 
            <code>
              multiple of 10
            </code>
          </div>
        </div>
      </details>
    </li>
  </ul>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  />
</details>
`;

exports[`JSONSchemaViewer - generateFriendlyName cases test OR / XOR / AND 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <ul>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items[0]
          </code>
          Â 
          <span>
            string OR number
          </span>
        </summary>
        <div>
          <span
            className="badge badge--info"
          >
            anyOf
          </span>
          <div
            className="tabs-container"
          >
            <ul
              aria-orientation="horizontal"
              className="tabs"
              role="tablist"
            >
              <li
                aria-selected={true}
                onClick={[Function]}
                tabIndex={0}
              >
                1
              </li>
              <li
                aria-selected={false}
                onClick={[Function]}
                tabIndex={-1}
              >
                2
              </li>
            </ul>
            <div>
              <div
                hidden={false}
                role="tabpanel"
              >
                <div
                  role="tabpanel"
                >
                  <strong>
                    type
                  </strong>
                  Â :Â 
                  <span
                    style={
                      {
                        "opacity": "0.6",
                      }
                    }
                  >
                    string
                  </span>
                  <div
                    style={
                      {
                        "marginTop": "var(--ifm-table-cell-padding)",
                      }
                    }
                  >
                    <div>
                      <strong>
                        Length :
                      </strong>
                      Â 
                      <code>
                        &lt;= 5 character(s)
                      </code>
                    </div>
                  </div>
                </div>
              </div>
              <div
                hidden={true}
                role="tabpanel"
              >
                <div
                  role="tabpanel"
                >
                  <strong>
                    type
                  </strong>
                  Â :Â 
                  <span
                    style={
                      {
                        "opacity": "0.6",
                      }
                    }
                  >
                    number
                  </span>
                  <div
                    style={
                      {
                        "marginTop": "var(--ifm-table-cell-padding)",
                      }
                    }
                  >
                    <div>
                      <strong>
                        Possible values :
                      </strong>
                      Â 
                      <code>
                        &gt;= 0
                      </code>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </details>
    </li>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items[1]
          </code>
          Â 
          <span>
            number
          </span>
        </summary>
        <div>
          <span
            className="badge badge--info"
          >
            oneOf
          </span>
          <div
            className="tabs-container"
          >
            <ul
              aria-orientation="horizontal"
              className="tabs"
              role="tablist"
            >
              <li
                aria-selected={true}
                onClick={[Function]}
                tabIndex={0}
              >
                1
              </li>
              <li
                aria-selected={false}
                onClick={[Function]}
                tabIndex={-1}
              >
                2
              </li>
            </ul>
            <div>
              <div
                hidden={false}
                role="tabpanel"
              >
                <div
                  role="tabpanel"
                >
                  <strong>
                    type
                  </strong>
                  Â :Â 
                  <span
                    style={
                      {
                        "opacity": "0.6",
                      }
                    }
                  >
                    number
                  </span>
                  <div
                    style={
                      {
                        "marginTop": "var(--ifm-table-cell-padding)",
                      }
                    }
                  >
                    <div>
                      <strong>
                        Possible values :
                      </strong>
                      Â 
                      <code>
                        multiple of 5
                      </code>
                    </div>
                  </div>
                </div>
              </div>
              <div
                hidden={true}
                role="tabpanel"
              >
                <div
                  role="tabpanel"
                >
                  <strong>
                    type
                  </strong>
                  Â :Â 
                  <span
                    style={
                      {
                        "opacity": "0.6",
                      }
                    }
                  >
                    number
                  </span>
                  <div
                    style={
                      {
                        "marginTop": "var(--ifm-table-cell-padding)",
                      }
                    }
                  >
                    <div>
                      <strong>
                        Possible values :
                      </strong>
                      Â 
                      <code>
                        multiple of 3
                      </code>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </details>
    </li>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items[2]
          </code>
          Â 
          <span>
            string
          </span>
        </summary>
        <div>
          <span
            className="badge badge--info"
          >
            allOf
          </span>
          <div
            className="tabs-container"
          >
            <ul
              aria-orientation="horizontal"
              className="tabs"
              role="tablist"
            >
              <li
                aria-selected={true}
                onClick={[Function]}
                tabIndex={0}
              >
                1
              </li>
              <li
                aria-selected={false}
                onClick={[Function]}
                tabIndex={-1}
              >
                2
              </li>
            </ul>
            <div>
              <div
                hidden={false}
                role="tabpanel"
              >
                <div
                  role="tabpanel"
                >
                  <strong>
                    type
                  </strong>
                  Â :Â 
                  <span
                    style={
                      {
                        "opacity": "0.6",
                      }
                    }
                  >
                    string
                  </span>
                  <div
                    style={
                      {
                        "marginTop": "var(--ifm-table-cell-padding)",
                      }
                    }
                  />
                </div>
              </div>
              <div
                hidden={true}
                role="tabpanel"
              >
                <div
                  role="tabpanel"
                >
                  <strong>
                    type
                  </strong>
                  Â :Â 
                  <span
                    style={
                      {
                        "opacity": "0.6",
                      }
                    }
                  >
                    string
                  </span>
                  <div
                    style={
                      {
                        "marginTop": "var(--ifm-table-cell-padding)",
                      }
                    }
                  >
                    <div>
                      <strong>
                        Length :
                      </strong>
                      Â 
                      <code>
                        &lt;= 5 character(s)
                      </code>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </details>
    </li>
  </ul>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  />
</details>
`;

exports[`JSONSchemaViewer - generateFriendlyName cases test Object without explicit type 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <ul>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items[0]
          </code>
          Â 
          <span>
            object
          </span>
        </summary>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style={
            {
              "opacity": "0.6",
            }
          }
        >
          object
        </span>
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        >
          <div>
            <strong>
              Length :
            </strong>
            Â 
            <code>
              &gt;= 3 propertie(s)
            </code>
          </div>
        </div>
      </details>
    </li>
  </ul>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  />
</details>
`;

exports[`JSONSchemaViewer - generateFriendlyName cases test Prefer title 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <ul>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items[0]
          </code>
          Â 
          <span>
            My custom title
          </span>
        </summary>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style={
            {
              "opacity": "0.6",
            }
          }
        >
          object
        </span>
        <ul>
          <li>
            <details
              open={false}
            >
              <summary>
                <strong>
                  text
                </strong>
                Â 
                <span>
                  string
                </span>
              </summary>
              <strong>
                type
              </strong>
              Â :Â 
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                string
              </span>
              <div
                style={
                  {
                    "marginTop": "var(--ifm-table-cell-padding)",
                  }
                }
              >
                <div>
                  <strong>
                    Constant value :
                  </strong>
                  Â 
                  <code>
                    Hello World
                  </code>
                </div>
              </div>
            </details>
          </li>
        </ul>
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        />
      </details>
    </li>
  </ul>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  />
</details>
`;

exports[`JSONSchemaViewer - generateFriendlyName cases test Simple array 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <ul>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items[0]
          </code>
          Â 
          <span>
            array
          </span>
        </summary>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style={
            {
              "opacity": "0.6",
            }
          }
        >
          array
        </span>
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        />
      </details>
    </li>
  </ul>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  />
</details>
`;
